File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        int n=arr.length;
        int[] left=new int[n];
        int[] right=new int[n];
        Stack<Integer> st=new Stack<>();
        for (int i=0;i<n;i++){
            while(!st.isEmpty() && arr[st.peek()]>arr[i]) st.pop();
            left[i]=st.isEmpty()?(i+1):(i-st.peek());
            st.push(i);
        }
        st.clear();
        for (int i=n-1;i>=0;i--){
            while(!st.isEmpty() && arr[st.peek()]>=arr[i]) st.pop();
            right[i]=st.isEmpty()?(n-i):(st.peek()-i);
            st.push(i);
        }
        long result=0;
        long mod=1_000_000_007L;
        for (int i=0;i<n;i++){
            result=(result+(long)arr[i]*left[i]%mod*right[i])%mod;
        }
        return (int)result;
    }
}

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        n=len(arr)
        mod=10**9+7
        left=[0]*n
        right=[0]*n
        st=[]
        for i in range(n):
            while st and arr[st[-1]]>arr[i]:
                st.pop()
            left[i]=i-(st[-1]if st else -1)
            st.append(i)
        st=[]
        for i in range(n-1,-1,-1):
            while st and arr[st[-1]]>=arr[i]:
                st.pop()
            right[i]=(st[-1] if st else n)-i
            st.append(i)
        result=0
        for i in range(n):
            result+=arr[i]*left[i]*right[i]
        return result%mod